File: apps/web/src/lib/api/schemas/concurrency.ts
import { z } from "zod";

const concurrencyLimitSchema = z.object({
  id: z.string(),
  workspace_id: z.string(),
  name: z.string(),
  limit: z.number(),
  in_flight: z.number().default(0),
  resource_type: z.string(),
  resource_id: z.string().nullish(),
  available: z.number(),
  saturated: z.boolean(),
  created_at: z.string(),
  updated_at: z.string(),
});
export type ConcurrencyLimit = z.infer<typeof concurrencyLimitSchema>;

export const concurrencyLimitListSchema = z.object({
  limits: z.array(concurrencyLimitSchema).default([]),
});
export type ConcurrencyLimitList = z.infer<typeof concurrencyLimitListSchema>;

File: packages/shared/src/shared/http/concurrency.py
from __future__ import annotations

from datetime import datetime

from pydantic import Field, JsonValue

from shared.app_identity import DEFAULT_RESOURCE_TYPE
from shared.http.base import HttpModel


class ConcurrencyLimitSetRequest(HttpModel):
    name: str
    limit: int
    workspace: str = "default"
    resource_type: str = DEFAULT_RESOURCE_TYPE
    resource_id: str | None = None
    metadata: dict[str, JsonValue] = Field(default_factory=dict)


class ConcurrencyLimitResponse(HttpModel):
    id: str
    workspace_id: str
    name: str
    limit: int
    in_flight: int = 0
    resource_type: str = DEFAULT_RESOURCE_TYPE
    resource_id: str | None = None
    metadata: dict[str, JsonValue] = Field(default_factory=dict)
    available: int
    saturated: bool
    created_at: datetime
    updated_at: datetime


class ConcurrencyLimitListResponse(HttpModel):
    limits: list[ConcurrencyLimitResponse] = Field(default_factory=list)


class ConcurrencyAcquireResponse(HttpModel):
    status: str
    acquired: bool
    record: ConcurrencyLimitResponse
    available_before: int
    available_after: int
    reason: str


__all__ = [
    "ConcurrencyAcquireResponse",
    "ConcurrencyLimitListResponse",
    "ConcurrencyLimitResponse",
    "ConcurrencyLimitSetRequest",
]
